File: bot/trading_bot.py
"""
Main trading bot that orchestrates detection, simulation, and execution
"""
import asyncio
import logging
from typing import Dict, Any, Optional
from datetime import datetime
import bittensor as bt
from bot.config import config
from bot.mempool_listener import MempoolListener
from bot.price_simulator import PriceSimulator, SubnetPool
from bot.execution_engine import ExecutionEngine
from bot.database import AsyncSessionLocal
from bot.models import Trade, SubnetPool as SubnetPoolModel

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class TradingBot:
    """Main trading bot - optimized for minimal latency"""
    
    def __init__(self):
        self.mempool_listener = MempoolListener()
        self.execution_engine = ExecutionEngine()
        self.subtensor = bt.subtensor(network="finney")
        self.running = False
        self.daily_trades = 0
        self.last_reset = datetime.now().date()
        # In-memory pool cache for ultra-fast access
        self.pool_cache: Dict[int, tuple] = {}  # netuid -> (SubnetPool, timestamp)
        self.cache_ttl = 10  # Cache for 10 seconds
        
        # Latency tracking
        self.latency_metrics = {
            "detect_to_decision": [],
            "decision_to_execute": [],
            "total_pipeline": []
        }
    
    async def initialize(self):
        """Initialize bot components"""
        logger.info("Initializing trading bot...")
        
        # Initialize execution engine
        await self.execution_engine.initialize()
        
        # Register mempool callback
        self.mempool_listener.register_callback(self._handle_stake_detection)
        
        # Initialize database
        from bot.database import init_db
        await init_db()
        
        # Start background liquidity updater
        asyncio.create_task(self._background_liquidity_updater())
        
        logger.info("Trading bot initialized")
    
    async def _background_liquidity_updater(self):
        """Background task to update liquidity cache every block (~12 seconds)"""
        last_block = 0
        
        while self.running:
            try:
                # Get current block
                loop = asyncio.get_event_loop()
                current_block = await loop.run_in_executor(
                    None,
                    lambda: self.subtensor.get_current_block()
                )
                
                # Update cache if new block
                if current_block > last_block:
                    last_block = current_block
                    
                    # Update liquidity for all monitored subnets in parallel
                    tasks = [
                        self._get_subnet_pool(netuid) 
                        for netuid in config.monitored_subnets
                    ]
                    await asyncio.gather(*tasks, return_exceptions=True)
                    
                    logger.debug(f"Liquidity cache updated at block {current_block}")
                
                # Check every 2 seconds (faster than block time for responsiveness)
                await asyncio.sleep(2)
                
            except Exception as e:
                logger.debug(f"Liquidity updater error: {e}")
                await asyncio.sleep(5)
    
    async def _get_subnet_pool(self, netuid: int) -> Optional[SubnetPool]:
        """Get current subnet pool state - optimized with in-memory cache"""
        import time
        
        # Check in-memory cache first (fastest)
        if netuid in self.pool_cache:
            pool, cache_time = self.pool_cache[netuid]
            if time.time() - cache_time < self.cache_ttl:
                return pool
        
        # Cache miss - fetch fresh data (async, non-blocking)
        try:
            # Use thread pool for blocking subtensor calls
            loop = asyncio.get_event_loop()
            
            def _fetch_pool():
                try:
                    subnet_info = self.subtensor.subnet_info(netuid=netuid)
                    tao_reserve = float(subnet_info.get('tao_in', 1000))
                    alpha_reserve = float(subnet_info.get('alpha_in', 500))
                    return tao_reserve, alpha_reserve
                except Exception as e:
                    logger.warning(f"Could not fetch subnet info: {e}, using defaults")
                    # Fast fallback
                    return 1000.0, 500.0
            
            tao_reserve, alpha_reserve = await loop.run_in_executor(None, _fetch_pool)
            
            pool = SubnetPool(tao_reserve, alpha_reserve)
            
            # Update in-memory cache
            self.pool_cache[netuid] = (pool, time.time())
            
            # Update DB cache in background (non-blocking)
            asyncio.create_task(self._update_db_cache(netuid, tao_reserve, alpha_reserve))
            
            return pool
            
        except Exception as e:
            logger.error(f"Error getting subnet pool: {e}")
            return None
    
    async def _update_db_cache(self, netuid: int, tao_reserve: float, alpha_reserve: float):
        """Update database cache asynchronously"""
        try:
            async with AsyncSessionLocal() as session:
                pool_model = SubnetPoolModel(
                    subnet_id=netuid,
                    tao_reserve=tao_reserve,
                    alpha_reserve=alpha_reserve,
                    current_price=tao_reserve / alpha_reserve if alpha_reserve > 0 else 0
                )
                session.merge(pool_model)
                await session.commit()
        except Exception as e:
            logger.debug(f"Error updating DB cache: {e}")
    
    async def _handle_stake_detection(self, tx_data: Dict[str, Any]):
        """Handle detected stake transaction - optimized for speed with latency tracking"""
        import time
        pipeline_start = time.time()
        detect_time = tx_data.get("timestamp", pipeline_start)
        
        if tx_data.get("type") != "stake":
            return
        
        # Reset daily counter if needed
        if datetime.now().date() > self.last_reset:
            self.daily_trades = 0
            self.last_reset = datetime.now().date()
        
        # Check daily limit (fast check, no I/O)
        if self.daily_trades >= config.max_daily_trades:
            return
        
        netuid = tx_data["netuid"]
        wallet_stake = tx_data["amount"]
        wallet_address = tx_data.get("hotkey_ss58", "unknown")
        
        # Ultra-fast filter: quick threshold check before any computation
        if wallet_stake < config.min_wallet_stake:
            return
        
        logger.info(f"⚡ DETECTED: {wallet_stake} TAO stake on subnet {netuid}")
        
        decision_start = time.time()
        
        # Get pool (cached, should be <1ms)
        pool = await self._get_subnet_pool(netuid)
        if not pool:
            logger.warning(f"Could not get pool state for subnet {netuid}")
            return
        
        # Fast simulation (in-memory, no I/O) - target <5ms
        result = PriceSimulator.find_optimal_stake(
            pool=pool,
            wallet_stake=wallet_stake,
            max_bot_stake=config.max_bot_stake,
            min_profit=config.min_expected_profit
        )
        
        decision_time = time.time() - decision_start
        self._record_latency("detect_to_decision", decision_time * 1000)
        
        if not result:
            return
        
        optimal_stake, expected_profit, price_move = result
        
        logger.info(
            f"✅ PROFITABLE! Bot: {optimal_stake} TAO, "
            f"Profit: {expected_profit:.4f} TAO, "
            f"Move: {price_move:.2f}%, "
            f"Decision: {decision_time*1000:.1f}ms"
        )
        
        # Create trade data
        trade_id = f"{netuid}_{wallet_address}_{time.time()}"
        
        trade_data = {
            "subnet_id": netuid,
            "wallet_address": wallet_address,
            "wallet_stake": wallet_stake,
            "bot_stake": optimal_stake,
            "price_before": pool.price(),
            "expected_profit": expected_profit,
            "status": "pending",
            "wallet_tx": tx_data.get("tx_hash"),
            "detect_timestamp": detect_time,
            "decision_timestamp": decision_start
        }
        
        # Execute trade immediately (don't await - fire and continue)
        execute_start = time.time()
        asyncio.create_task(self._execute_trade(trade_id, trade_data, pool, execute_start))
        
        # Record total pipeline latency
        total_time = time.time() - pipeline_start
        self._record_latency("total_pipeline", total_time * 1000)
    
    def _record_latency(self, metric: str, latency_ms: float):
        """Record latency metric"""
        if metric in self.latency_metrics:
            self.latency_metrics[metric].append(latency_ms)
            # Keep only last 100 measurements
            if len(self.latency_metrics[metric]) > 100:
                self.latency_metrics[metric] = self.latency_metrics[metric][-100:]
            
            # Log statistics periodically
            if len(self.latency_metrics[metric]) % 10 == 0:
                metrics = self.latency_metrics[metric]
                avg = sum(metrics) / len(metrics)
                min_lat = min(metrics)
                max_lat = max(metrics)
                logger.info(
                    f"📊 {metric}: avg={avg:.1f}ms, min={min_lat:.1f}ms, max={max_lat:.1f}ms"
                )
    
    async def _execute_trade(
        self,
        trade_id: str,
        trade_data: Dict[str, Any],
        pool: SubnetPool,
        execute_start: float
    ):
        """Execute front-run trade with latency tracking"""
        import time
        try:
            netuid = trade_data["subnet_id"]
            bot_stake = trade_data["bot_stake"]
            
            # Step 1: Bot stakes
            logger.info(f"Executing bot stake: {bot_stake} TAO on subnet {netuid}")
            stake_start = time.time()
            stake_tx = await self.execution_engine.execute_stake(
                netuid=netuid,
                amount=bot_stake,
                trade_id=trade_id
            )
            
            execute_time = time.time() - execute_start
            self._record_latency("decision_to_execute", execute_time * 1000)
            
            if not stake_tx:
                logger.error("Bot stake failed")
                trade_data["status"] = "failed"
                trade_data["error_message"] = "Bot stake transaction failed"
                await self._record_trade(trade_data)
                return
            
            trade_data["bot_stake_tx"] = stake_tx
            trade_data["status"] = "staked"
            
            # Wait for wallet stake to execute (monitor block)
            logger.info("Waiting for wallet stake to execute...")
            await asyncio.sleep(12)  # Approximate block time
            
            # Step 2: Bot unstakes
            logger.info(f"Executing bot unstake: {bot_stake} TAO from subnet {netuid}")
            unstake_tx = await self.execution_engine.execute_unstake(
                netuid=netuid,
                amount=bot_stake,
                trade_id=trade_id
            )
            
            if not unstake_tx:
                logger.error("Bot unstake failed")
                trade_data["status"] = "failed"
                trade_data["error_message"] = "Bot unstake transaction failed"
            else:
                trade_data["bot_unstake_tx"] = unstake_tx
                trade_data["status"] = "completed"
                
                # Calculate actual profit (simplified)
                # In production, you'd query the actual balance change
                trade_data["actual_profit"] = trade_data["expected_profit"]
            
            # Record trade asynchronously (non-blocking)
            await self.execution_engine.record_trade_async(trade_data)
            self.daily_trades += 1
            
            logger.info(f"✅ Trade completed: {trade_id}")
            
        except Exception as e:
            logger.error(f"Error executing trade: {e}")
            trade_data["status"] = "failed"
            trade_data["error_message"] = str(e)
            await self._record_trade(trade_data)
    
    async def _record_trade(self, trade_data: Dict[str, Any]):
        """Record trade in database"""
        try:
            async with AsyncSessionLocal() as session:
                trade = Trade(**trade_data)
                session.add(trade)
                await session.commit()
        except Exception as e:
            logger.error(f"Error recording trade: {e}")
    
    async def start(self):
        """Start the trading bot"""
        self.running = True
        logger.info("Starting trading bot...")
        
        # Start mempool listener
        await self.mempool_listener.start()
    
    def get_latency_stats(self) -> Dict[str, Dict[str, float]]:
        """Get latency statistics"""
        stats = {}
        for metric, values in self.latency_metrics.items():
            if values:
                stats[metric] = {
                    "avg": sum(values) / len(values),
                    "min": min(values),
                    "max": max(values),
                    "p50": sorted(values)[len(values) // 2],
                    "p95": sorted(values)[int(len(values) * 0.95)] if len(values) > 1 else values[0],
                    "count": len(values)
                }
        return stats
    
    async def stop(self):
        """Stop the trading bot"""
        self.running = False
        await self.mempool_listener.stop()
        
        # Print final latency statistics
        stats = self.get_latency_stats()
        if stats:
            logger.info("📊 Final Latency Statistics:")
            for metric, stat in stats.items():
                logger.info(
                    f"  {metric}: avg={stat['avg']:.1f}ms, "
                    f"min={stat['min']:.1f}ms, max={stat['max']:.1f}ms, "
                    f"p95={stat['p95']:.1f}ms"
                )
        
        logger.info("Trading bot stopped")


async def main():
    """Main entry point"""
    bot = TradingBot()
    await bot.initialize()
    
    try:
        await bot.start()
    except KeyboardInterrupt:
        logger.info("Shutting down...")
        await bot.stop()


if __name__ == "__main__":
    asyncio.run(main())

File: src/trading_bot.rs
use anyhow::Result;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;
use tokio::time::sleep;
use tracing::{debug, error, info};
use subxt::OnlineClient;
use subxt::config::PolkadotConfig;
use crate::config;
use crate::mempool_listener::{MempoolListener, StakeData};
use crate::price_simulator::{PriceSimulator, SubnetPool};
use crate::execution_engine::ExecutionEngine;
use crate::models::Trade;
use crate::database;

pub struct TradingBot {
    mempool_listener: MempoolListener,
    execution_engine: ExecutionEngine,
    subtensor: Arc<RwLock<Option<OnlineClient<PolkadotConfig>>>>,
    running: Arc<RwLock<bool>>,
    daily_trades: Arc<RwLock<i32>>,
    pool_cache: Arc<RwLock<HashMap<i32, (SubnetPool, f64)>>>,
}

impl TradingBot {
    pub fn new() -> Self {
        Self {
            mempool_listener: MempoolListener::new(),
            execution_engine: ExecutionEngine::new(),
            subtensor: Arc::new(RwLock::new(None)),
            running: Arc::new(RwLock::new(false)),
            daily_trades: Arc::new(RwLock::new(0)),
            pool_cache: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    pub async fn initialize(&self) -> Result<()> {
        info!("Initializing trading bot...");

        // Initialize subtensor connection
        let config = config::get_config();
        let config_guard = config.read().await;
        let rpc_url = &config_guard.subtensor_rpc;

        info!("Connecting to Bittensor network (this may take a moment)...");
        match OnlineClient::from_url(rpc_url).await {
            Ok(client) => {
                *self.subtensor.write().await = Some(client);
                info!("Successfully connected to Bittensor network");
            }
            Err(e) => {
                error!("Failed to initialize subtensor: {}", e);
                return Err(e.into());
            }
        }

        // Initialize execution engine
        self.execution_engine.initialize().await?;

        // Register mempool callback
        let bot = self.clone();
        self.mempool_listener.register_callback(Arc::new(move |stake_data| {
            let bot_clone = bot.clone();
            tokio::spawn(async move {
                bot_clone.handle_stake_detection(stake_data).await;
            });
        })).await;

        // Initialize database
        database::init_db().await?;

        info!("Trading bot initialized");
        Ok(())
    }

    async fn get_subnet_pool(&self, netuid: i32) -> SubnetPool {
        // Check in-memory cache first
        {
            let cache = self.pool_cache.read().await;
            if let Some((pool, cache_time)) = cache.get(&netuid) {
                let now = std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap()
                    .as_secs_f64();
                if now - cache_time < 10.0 {
                    return pool.clone();
                }
            }
        }

        // Cache miss - fetch fresh data
        // Placeholder - implement actual pool fetching from chain
        let pool = SubnetPool::new(1000.0, 500.0);
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs_f64();
        
        self.pool_cache.write().await.insert(netuid, (pool.clone(), now));
        pool
    }

    async fn handle_stake_detection(&self, tx_data: StakeData) {
        let netuid = tx_data.netuid;
        let wallet_stake = tx_data.amount;
        let wallet_address = tx_data.hotkey_ss58.clone();

        // Get bot wallet balance
        let bot_balance = self.execution_engine.get_wallet_balance().await;

        if bot_balance.is_none() {
            info!("Could not get bot wallet balance");
            return;
        }

        let bot_balance = bot_balance.unwrap();
        let config = config::get_config();
        let config_guard = config.read().await;
        let min_reserve = config_guard.min_wallet_reserve;

        // Calculate stake amount
        let bot_stake = if bot_balance >= wallet_stake + min_reserve {
            wallet_stake
        } else {
            (bot_balance - min_reserve).max(0.0)
        };

        if bot_stake <= 0.0 {
            info!(
                "⏭️ Skipping trade: insufficient balance (balance: {:.4} TAO, need: {:.4} TAO, reserve: {:.4} TAO)",
                bot_balance, wallet_stake, min_reserve
            );
            return;
        }

        info!(
            "✅ Staking: {:.4} TAO (wallet stake: {:.4} TAO, bot balance: {:.4} TAO)",
            bot_stake, wallet_stake, bot_balance
        );

        // Create trade data
        let trade_id = format!("{}_{}_{}", netuid, wallet_address, 
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs());

        let trade_data = Trade {
            id: None,
            timestamp: chrono::Utc::now(),
            subnet_id: netuid,
            wallet_address,
            wallet_stake,
            bot_stake,
            price_after: None,
            actual_profit: None,
            bot_stake_tx: None,
            bot_unstake_tx: None,
            wallet_tx: None,
            status: "pending".to_string(),
            error_message: None,
        };

        // Execute trade immediately (fire and continue)
        let bot = self.clone();
        tokio::spawn(async move {
            bot.execute_trade(trade_id, trade_data).await;
        });
    }

    async fn execute_trade(&self, trade_id: String, mut trade_data: Trade) {
        let netuid = trade_data.subnet_id;
        let bot_stake = trade_data.bot_stake;

        // Step 1: Bot stakes
        info!("Executing bot stake: {:.4} TAO on subnet {}", bot_stake, netuid);
        
        match self.execution_engine.execute_stake(netuid, bot_stake, Some(trade_id.clone())).await {
            Ok(Some(stake_tx)) => {
                trade_data.bot_stake_tx = Some(stake_tx);
                trade_data.status = "staked".to_string();
            }
            Ok(None) => {
                error!("Bot stake failed");
                trade_data.status = "failed".to_string();
                trade_data.error_message = Some("Bot stake transaction failed".to_string());
                self.record_trade(trade_data).await;
                return;
            }
            Err(e) => {
                error!("Bot stake error: {}", e);
                trade_data.status = "failed".to_string();
                trade_data.error_message = Some(format!("Bot stake error: {}", e));
                self.record_trade(trade_data).await;
                return;
            }
        }

        // Wait for wallet stake to execute (monitor block)
        info!("Waiting for wallet stake to execute...");
        sleep(tokio::time::Duration::from_secs(12)).await;

        // Step 2: Bot unstakes
        info!("Executing bot unstake: {:.4} TAO from subnet {}", bot_stake, netuid);
        
        match self.execution_engine.execute_unstake(netuid, bot_stake, Some(trade_id.clone())).await {
            Ok(Some(unstake_tx)) => {
                trade_data.bot_unstake_tx = Some(unstake_tx);
                trade_data.status = "completed".to_string();
            }
            Ok(None) => {
                error!("Bot unstake failed");
                trade_data.status = "failed".to_string();
                trade_data.error_message = Some("Bot unstake transaction failed".to_string());
            }
            Err(e) => {
                error!("Bot unstake error: {}", e);
                trade_data.status = "failed".to_string();
                trade_data.error_message = Some(format!("Bot unstake error: {}", e));
            }
        }

        // Record trade asynchronously
        self.execution_engine.record_trade_async(trade_data.clone()).await;
        *self.daily_trades.write().await += 1;

        info!("✅ Trade completed: {}", trade_id);
    }

    async fn record_trade(&self, trade_data: Trade) {
        // Record trade in database
        // This would use sqlx to insert
    }

    pub async fn start(&self) -> Result<()> {
        *self.running.write().await = true;
        info!("Starting trading bot...");

        // Start mempool listener
        let listener = self.mempool_listener.clone();
        tokio::spawn(async move {
            if let Err(e) = listener.start().await {
                error!("Mempool listener error: {}", e);
            }
        });

        // Keep main thread alive
        loop {
            sleep(tokio::time::Duration::from_secs(1)).await;
            if !*self.running.read().await {
                break;
            }
        }

        Ok(())
    }

    pub async fn stop(&self) {
        *self.running.write().await = false;
        self.mempool_listener.stop().await;
        info!("Trading bot stopped");
    }
}

impl Clone for TradingBot {
    fn clone(&self) -> Self {
        Self {
            mempool_listener: self.mempool_listener.clone(),
            execution_engine: self.execution_engine.clone(),
            subtensor: Arc::clone(&self.subtensor),
            running: Arc::clone(&self.running),
            daily_trades: Arc::clone(&self.daily_trades),
            pool_cache: Arc::clone(&self.pool_cache),
        }
    }
}
